test/numerical/TestGRU.cpp

```diff
@@ -24,10 +24,28 @@ bool isOMGRUTheSameAsNaiveImplFor(const int direction, const int S, const int B,
     const int I, const int H, const int linearBeforeReset,
     bool isDynamicS = false, bool isDynamicB = false) {
 
+  static int testNum = 0;
+  printf("attempt %d with direction %d, S %d, B %d, I %d, H %d, "
+         "linearBeforeReset %d, isDynS %d, isDynB %d\n",
+      ++testNum, direction, S, B, I, H, linearBeforeReset, isDynamicS,
+      isDynamicB);
   GRULibBuilder gru(SHARED_LIB_BASE.str(), direction, S, B, I, H,
       linearBeforeReset, isDynamicS, isDynamicB);
-  return gru.build() && gru.compileAndLoad() && gru.prepareInputs() &&
-         gru.run() && gru.verifyOutputs();
+  bool successBuild = gru.build() && gru.compileAndLoad();
+  if (!successBuild) {
+    printf("Fail to build.\n");
+    return false;
+  }
+  // Verify generated library
+  std::string instructionName =
+      getenv("TEST_CHECK_INSTRUCTION") ? getenv("TEST_CHECK_INSTRUCTION") : "";
+  std::string sharedLibName =
+      ModelLibBuilder::getSharedLibName(SHARED_LIB_BASE.str());
+  if (!ModelLibBuilder::checkSharedLibInstruction(
+          instructionName, sharedLibName))
+    return false;
```

Comment:
Can you refactor this code to be much more concise.

```
return gru.build() && gru.compileAndLoad() &&  gru.checkInstruction("TEST_CHECK_INSTRUCTION") && 
          gru.prepareInputs() && gru.run() && gru.verifyOutputs();
```

then you can create a `checkInstruction` that read the env var, if empty return true, and if not empty return ` checkSharedLibInstruction( instructionName, sharedLibName))`. Note that the lib name is already stored in the parent class, which you see when you created `gru` object, so no need to pass it.

If you want to do 2 versions from checkInstruction, one that that the instruction, and another named `checkInstructionFromEnv` and that simply check for the env var, and if undefined return true, otherwise call the `checkInstruction` where the actual instruction string is passed, that would be even cleaner.